src/macrokernels.jl

```diff
@@ -1,84 +1,85 @@
-#         Cₘₙ = zero(eltype(C))
-#         for k ∈ CloseOpen(K)
-#             Cₘₙ += A[m,k] * B[k,n]
-#         end
-#         C[m,n] += α * Cₘₙ
-#     end
-#     nothing
-# end
+@inline incrementp(A::AbstractStridedPointer{<:Base.HWReal,3}, a::Ptr) = VectorizationBase.increment_ptr(A, a, (Zero(), Zero(), One()))
+@inline increment2(B::AbstractStridedPointer{<:Base.HWReal,2}, b::Ptr, ::StaticInt{nᵣ}) where {nᵣ} = VectorizationBase.increment_ptr(B, b, (Zero(), StaticInt{nᵣ}()))
+@inline increment1(C::AbstractStridedPointer{<:Base.HWReal,2}, c::Ptr, ::StaticInt{mᵣW}) where {mᵣW} = VectorizationBase.increment_ptr(C, c, (StaticInt{mᵣW}(), Zero()))
+macro kernel(pack::Bool, ex::Expr)
+  ex.head === :for || throw(ArgumentError("Must be a matmul for loop."))
+  mincrements = Expr[:(c = increment1(C, c, mᵣW)), :(ãₚ = incrementp(Ãₚ, ãₚ)), :(m = vsub_nsw(m, mᵣW))]
+  # massumes = Expr[:(assume(m < mᵣW)),
+  #                 :(assume(VectorizationBase.vgt(ãₚ, VectorizationBase.increment_ptr($(esc(:Ãₚ)), ãₚ, (vsub_nsw($(esc(:M)), mᵣW), LoopVectorization.Zero())), $(esc(:Ãₚ))))),
+  #                 :(assume(VectorizationBase.vgt(c, VectorizationBase.increment_ptr($(esc(:C)), c, (vsub_nsw($(esc(:M)), mᵣW), LoopVectorization.Zero())), $(esc(:C)))))]
+  offsetprecalc = GlobalRef(VectorizationBase,:offsetprecalc)
+  preheader = quote
+    mᵣ, nᵣ = matmul_params(Val($(esc(:T))))
+    mᵣW = pick_vector_width($(esc(:T))) * mᵣ
+    m = $(esc(:M)) % Int32
+    n = $(esc(:N)) % Int32
+    Ãₚ = $(esc(:Ãₚ))
+    B = $offsetprecalc($(esc(:B)), Val{(9,9)}())
+    C = $offsetprecalc($(esc(:C)), Val{(9,9)}())
+    b = pointer(B); c = pointer(C); ãₚ = pointer(Ãₚ)
+  end
+  if pack
+    push!(mincrements, :(a = increment1(A, a, mᵣW)))
+    push!(preheader.args, :(A = $(esc(:A))), :(a = pointer(A)))
+    areconstruct = Expr[:($(esc(:A)) = VectorizationBase.reconstruct_ptr(A, a))]
+    # push!(massumes, :(assume(VectorizationBase.vgt(a, VectorizationBase.increment_ptr($(esc(:A)), a, (vsub_nsw($(esc(:M)), mᵣW), LoopVectorization.Zero())), $(esc(:A))))))
+  else
+    Ainit = areconstruct = Expr[]
+  end
+  lvkern = esc(:(@avx inline=true $ex))
+
+  loopnest = quote
+    let ãₚ = ãₚ, c = c, $(esc(:B)) = VectorizationBase.reconstruct_ptr(B, b), m = m
+      while m ≥ mᵣW#VectorizationBase.vle(a, amax, A)
+        let $(esc(:M)) = mᵣW, $(esc(:N)) = nᵣ, $(esc(:Ãₚ)) = VectorizationBase.reconstruct_ptr(droplastdim(Ãₚ), ãₚ), $(esc(:C)) = VectorizationBase.reconstruct_ptr(C, c), $(areconstruct...)
+          $lvkern
+          $(mincrements...)
+        end
+      end
+      if m > zero(Int32)#vne(a, amax, A)
+        let $(esc(:M)) = UpperBoundedInteger((m%UInt)%Int, mᵣW - One()), $(esc(:N)) = nᵣ, $(esc(:Ãₚ)) = VectorizationBase.reconstruct_ptr(droplastdim(Ãₚ), ãₚ), $(esc(:C)) = VectorizationBase.reconstruct_ptr(C, c), $(areconstruct...)
+          # $(massumes...)
+          $lvkern
+        end
+      end
+    end
+  end
+
+  if !pack
```

Comment:
Are you sure it's `if !pack`?

Reply:
Yes.
`pack` means it is packing `Apack`, and `!pack` that it's using `Apack`.
When `pack`, it only evaluates a single `n_r` tile of `Bpack`, while with `!pack`, it needs to evaluate all of those remaining.

Reply:
Ah, I see.